src/hotspot/share/runtime/handshake.hpp

```diff
@@ -39,74 +40,75 @@ class JavaThread;
+   AsynchHandshakeClosure(const char* name) : HandshakeClosure(name) {}
+   virtual ~AsynchHandshakeClosure() {}
+   virtual bool is_asynch()          { return true; }
+};
+
 class Handshake : public AllStatic {
  public:
   // Execution of handshake operation
-  static void execute(HandshakeClosure* hs_cl);
-  static bool execute(HandshakeClosure* hs_cl, JavaThread* target);
-  static bool execute_direct(HandshakeClosure* hs_cl, JavaThread* target);
+  static void execute(HandshakeClosure*       hs_cl);
+  static void execute(HandshakeClosure*       hs_cl, JavaThread* target);
+  static void execute(AsynchHandshakeClosure* hs_cl, JavaThread* target);
 };
 
 // The HandshakeState keeps track of an ongoing handshake for this JavaThread.
-// VMThread/Handshaker and JavaThread are serialized with semaphore _processing_sem
-// making sure the operation is only done by either VMThread/Handshaker on behalf
-// of the JavaThread or by the target JavaThread itself.
+// VMThread/Handshaker and JavaThread are serialized with _lock making sure the
+// operation is only done by either VMThread/Handshaker on behalf of the
+// JavaThread or by the target JavaThread itself.
 class HandshakeState {
   JavaThread* _handshakee;
-  HandshakeOperation* volatile _operation;
-  HandshakeOperation* volatile _operation_direct;
-
-  Semaphore _handshake_turn_sem;  // Used to serialize direct handshakes for this JavaThread.
-  Semaphore _processing_sem;
-  bool _thread_in_process_handshake;
+  FilterQueue<HandshakeOperation*> _queue;
+  Mutex _lock;
+  Thread* _active_handshaker;
 
-  bool claim_handshake(bool is_direct);
+  bool claim_handshake();
   bool possibly_can_process_handshake();
   bool can_process_handshake();
-  void clear_handshake(bool is_direct);
-
   void process_self_inner();
 
-public:
-  HandshakeState();
+ public:
+  HandshakeState(JavaThread* thread);
 
-  void set_handshakee(JavaThread* thread) { _handshakee = thread; }
+  void add_operation(HandshakeOperation* op);
+  HandshakeOperation* pop_for_self();
+  HandshakeOperation* pop_for_processor();
 
-  void set_operation(HandshakeOperation* op);
-  bool has_operation() const { return _operation != NULL || _operation_direct != NULL; }
-  bool has_specific_operation(bool is_direct) const {
-    return is_direct ? _operation_direct != NULL : _operation != NULL;
+  bool has_operation_for_processor();
+  bool has_operation() {
+    return !_queue.is_empty();
   }
-
-  void process_by_self() {
-    if (!_thread_in_process_handshake) {
-      AutoModifyRestore<bool> temporarily(_thread_in_process_handshake, true);
-      process_self_inner();
-    }
+  bool block_for_operation() {
+    return !_queue.is_empty() || _lock.is_locked();
   }
 
+  void process_by_self();
+
   enum ProcessResult {
-    _no_operation = 0,
-    _not_safe,
-    _state_busy,
-    _success,
-    _number_states
+      _no_operation = 0,
+      _not_safe,
+      _claim_failed,
+      _processed,
+      _succeed,
+      _number_states
```

Comment:
Why are these indented by 4 spaces instead of 2 spaces?

Reply:
Fixed